airflow/models/dag.py

```diff
@@ -3505,3 +3562,68 @@ def get_current_dag(cls) -> DAG | None:
             return cls._context_managed_dags[0]
         except IndexError:
             return None
+
+
+def _run_task(ti: TaskInstance, session):
+    """
+    Run a single task instance, and push result to Xcom for downstream tasks. Bypasses a lot of
+    extra steps used in `task.run` to keep our local running as fast as possible
+    This function is only meant for the `dag.test` function as a helper function.
+
+    Args:
+        ti: TaskInstance to run
+    """
+    log.info("*****************************************************")
+    if ti.map_index > 0:
+        log.info("Running task %s index %d", ti.task_id, ti.map_index)
+    else:
+        log.info("Running task %s", ti.task_id)
+    try:
+        ti._run_raw_task(session=session)
+        session.flush()
+        log.info("%s ran successfully!", ti.task_id)
+    except AirflowSkipException:
+        log.info("Task Skipped, continuing")
+    log.info("*****************************************************")
+
+
+def _get_or_create_dagrun(
+    dag: DAG,
+    conf: dict[Any, Any] | None,
+    start_date: datetime,
+    execution_date: datetime,
+    run_id: str,
+    session: Session,
+) -> DagRun:
+    """
+    Create a DAGRun, but only after clearing the previous instance of said dagrun to prevent collisions.
+    This function is only meant for the `dag.test` function as a helper function.
+    :param dag: Dag to be used to find dagrun
+    :param conf: configuration to pass to newly created dagrun
+    :param start_date: start date of new dagrun, defaults to execution_date
+    :param execution_date: execution_date for finding the dagrun
+    :param run_id: run_id to pass to new dagrun
+    :param session: sqlalchemy session
+    :return:
+    """
+    log.info("dagrun id:" + dag.dag_id)
+    dr: DagRun = (
+        session.query(DagRun)
+        .filter(DagRun.dag_id == dag.dag_id, DagRun.execution_date == execution_date)
+        .first()
+    )
+    if dr:
+        session.delete(dr)
+        session.commit()
+    dr = dag.create_dagrun(
+        state=DagRunState.RUNNING,
+        execution_date=execution_date,
+        run_id=run_id,
+        start_date=start_date or execution_date,
+        session=session,
+        conf=conf,  # type: ignore
+    )
+    session.add(dr)
+    session.flush()
```

Comment:
`dag.create_dagrun` already calls this https://github.com/apache/airflow/blob/b885b906674d31fc2d15443e0d9d076007a170b1/airflow/models/dag.py#L2584-L2585)

Reply:
Ah good catch, yeah we can delete that.